shutdown/signal/signal.go

```diff
@@ -0,0 +1,45 @@
+package signal
+
+import (
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+)
+
+type Handler struct {
+	signalChannel chan os.Signal
+	wg            sync.WaitGroup
+}
+
+type Listener interface {
+	ShutdownSignalReceived()
+}
+
+func NewDefaultSignalHandler(listener Listener) *Handler {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
+	return NewSignalHandler(signals, listener)
+}
+
+func NewSignalHandler(signalChannel chan os.Signal, listener Listener) *Handler {
+	handler := Handler{signalChannel: signalChannel}
+
+	go func() {
+		handler.waitForSignalAndCallListener(signalChannel, listener)
+	}()
+
+	return &handler
+}
+
+func (h *Handler) waitForSignalAndCallListener(signalChannel chan os.Signal, listener Listener) {
+	h.wg.Add(1)
```

Comment:
afaik adding to the wg should be done in the same thread where one waits for it being done

-> should be usually in the main thread and not in the spawned go routine

Reply:
Hm. This would mean in this case that we have to replace the wait group with a channel that gets closed. Because multiple Threads might want to wait for the ShutdownHandler to complete.